google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/flexible/AppEngineFlexibleDeploymentEditor.java

```diff
@@ -147,116 +129,92 @@ protected void textChanged(DocumentEvent event) {
+    appYamlCombobox.setRenderer(new ListCellRendererWrapper<AppEngineFlexibleFacet>() {
       @Override
-      public void customize(JList list, Module value, int index, boolean selected,
+      public void customize(JList list, AppEngineFlexibleFacet facet, int index, boolean selected,
           boolean hasFocus) {
-        if (value != null) {
-          setText(value.getName());
+        if (facet != null) {
+          setText(tryTruncateConfigPathForDisplay(facet.getConfiguration().getAppYamlPath()));
         }
       }
     });
 
-    appYamlTextField.setText(getAppYamlPath());
+    editAppYamlButton.addActionListener(event -> openModuleSettings());
+    dockerfileDirectoryPathLink.addHyperlinkListener(event -> openModuleSettings());
+
+    updateSelectors();
+    toggleDockerfileSection();
+  }
+
+  private void openModuleSettings() {
+    AppEngineFlexibleFacet flexFacet = ((AppEngineFlexibleFacet) appYamlCombobox.getSelectedItem());
+
+    if (ModulesConfigurator.showFacetSettingsDialog(flexFacet, null /* tabNameToSelect */)) {
+      // The user may have updated the configuration, so we need to refresh it here too.
+      resetAppYamlConfigSelection(project);
+      toggleDockerfileSection();
 
-    moduleSettingsButton.addActionListener(event -> {
-      AppEngineFlexibleFacet flexFacet =
-          FacetManager.getInstance(((Module) modulesWithFlexFacetComboBox.getSelectedItem()))
-              .getFacetByType(AppEngineFlexibleFacetType.ID);
-      ModulesConfigurator.showFacetSettingsDialog(flexFacet, null /* tabNameToSelect */);
       // When we get out of the dialog window, we want to re-eval the configuration.
-      // validateConfiguration() can't be used here because the ConfigurationException isn't caught
-      // anywhere, and fireEditorStateChanged() doesn't trigger any listeners called from here.
-      // Emulating a user action triggers apply(), so that's what we're doing here.
+      // validateConfiguration() can't be used here because the ConfigurationException
+      // isn't caught anywhere, and fireEditorStateChanged() doesn't trigger any listeners
+      // called from here. Emulating a user action triggers apply(), so that's what we're
+      // doing here.
       hiddenValidationTrigger.doClick();
-      toggleDockerfileSection();
-      resetModuleConfigSelection(project);
-    });
+    }
+  }
 
-    updateSelectors();
-    toggleDockerfileSection();
+  private void resetRuntimeDisplay() {
+    if (isValidConfigurationFile(getAppYamlPath())) {
+      try {
+        Optional<FlexibleRuntime> flexibleRuntime =
+            appEngineProjectService.getFlexibleRuntimeFromAppYaml(getAppYamlPath());
+
+        runtimePanel.setLabelText(flexibleRuntime.map(FlexibleRuntime::toString).orElse(""));
+        runtimePanel.setVisible(true);
+        runtimePanel.setExplanationLabelVisibility(
+            !flexibleRuntime.map(FlexibleRuntime::isCustom).orElse(false));
```

Comment:
I had a mental todo to look into refactoring this if possible.

But the reason is that if you look a few lines up, I already extract the flexRuntime. I didn't want to call isCustomRuntime which would just duplicate this extraction (which requires IO - parsing the app.yaml file).